// @vendure/admin-ui/esm2022/core/providers/bulk-action-registry/bulk-action-registry.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class BulkActionRegistryService {
    constructor() {
        this.locationBulActionMap = new Map();
    }
    registerBulkAction(bulkAction) {
        if (!this.locationBulActionMap.has(bulkAction.location)) {
            this.locationBulActionMap.set(bulkAction.location, new Set([bulkAction]));
        }
        else {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            this.locationBulActionMap.get(bulkAction.location).add(bulkAction);
        }
    }
    getBulkActionsForLocation(id) {
        return [...(this.locationBulActionMap.get(id)?.values() ?? [])];
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: BulkActionRegistryService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: BulkActionRegistryService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: BulkActionRegistryService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
