// @vendure/admin-ui/esm2022/core/providers/localization/localization.service.mjs
import { Injectable } from '@angular/core';
import { map } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../../providers/i18n/i18n.service";
import * as i2 from "../../data/providers/data.service";
/**
 * @description
 * Provides localization helper functionality.
 *
 */
export class LocalizationService {
    constructor(i18nService, dataService) {
        this.i18nService = i18nService;
        this.dataService = dataService;
        this.uiLanguageAndLocale$ = this.dataService.client?.uiState()?.stream$?.pipe(map(({ uiState }) => {
            return [uiState.language, uiState.locale ?? undefined];
        }));
        this.direction$ = this.uiLanguageAndLocale$?.pipe(map(([languageCode]) => {
            return this.i18nService.isRTL(languageCode) ? 'rtl' : 'ltr';
        }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: LocalizationService, deps: [{ token: i1.I18nService }, { token: i2.DataService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: LocalizationService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: LocalizationService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [{ type: i1.I18nService }, { type: i2.DataService }] });
