// @vendure/admin-ui/esm2022/core/providers/data-table/data-table-config.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../local-storage/local-storage.service";
export class DataTableConfigService {
    constructor(localStorageService) {
        this.localStorageService = localStorageService;
    }
    getConfig(dataTableId) {
        const dataTableConfig = this.localStorageService.get('dataTableConfig') ?? {};
        if (dataTableId) {
            return (dataTableConfig[dataTableId] ?? {
                visibility: [],
                order: {},
                showSearchFilterRow: false,
                filterPresets: [],
            });
        }
        return dataTableConfig;
    }
    setConfig(idOrConfig, maybeConfig) {
        const currentConfig = this.getConfig();
        if (typeof idOrConfig === 'string') {
            if (maybeConfig) {
                this.localStorageService.set('dataTableConfig', {
                    ...currentConfig,
                    [idOrConfig]: maybeConfig,
                });
            }
        }
        else {
            this.localStorageService.set('dataTableConfig', { ...currentConfig, ...idOrConfig });
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DataTableConfigService, deps: [{ token: i1.LocalStorageService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DataTableConfigService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DataTableConfigService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [{ type: i1.LocalStorageService }] });
