// @vendure/admin-ui/esm2022/core/providers/data-table/data-table-filter-collection.mjs
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { assertNever } from '@vendure/common/lib/shared-utils';
import extend from 'just-extend';
import { Subject } from 'rxjs';
import { debounceTime, distinctUntilChanged, map, startWith, takeUntil } from 'rxjs/operators';
import { DataTableFilter, } from './data-table-filter';
export class FilterWithValue {
    constructor(filter, value, onUpdate) {
        this.filter = filter;
        this.value = value;
        this.onUpdateFns = new Set();
        if (onUpdate) {
            this.onUpdateFns.add(onUpdate);
        }
    }
    onUpdate(fn) {
        this.onUpdateFns.add(fn);
    }
    updateValue(value) {
        this.value = value;
        for (const fn of this.onUpdateFns) {
            fn(value);
        }
    }
    isId() {
        return this.filter.type.kind === 'id';
    }
    isText() {
        return this.filter.type.kind === 'text';
    }
    isNumber() {
        return this.filter.type.kind === 'number';
    }
    isBoolean() {
        return this.filter.type.kind === 'boolean';
    }
    isSelect() {
        return this.filter.type.kind === 'select';
    }
    isDateRange() {
        return this.filter.type.kind === 'dateRange';
    }
    isCustom() {
        return this.filter.type.kind === 'custom';
    }
}
export class DataTableFilterCollection {
    #filters;
    #activeFilters;
    #valueChanges$;
    #connectedToRouter;
    #filtersQueryParamName;
    constructor(router) {
        this.router = router;
        this.#filters = [];
        this.#activeFilters = [];
        this.#valueChanges$ = new Subject();
        this.#connectedToRouter = false;
        this.valueChanges = this.#valueChanges$.asObservable().pipe(debounceTime(10));
        this.#filtersQueryParamName = 'filters';
        this.destroy$ = new Subject();
    }
    get length() {
        return this.#filters.length;
    }
    get activeFilters() {
        return this.#activeFilters;
    }
    destroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    addFilter(config) {
        if (this.#connectedToRouter) {
            throw new Error('Cannot add filter after connecting to router. Make sure to call addFilter() before connectToRoute()');
        }
        this.#filters.push(new DataTableFilter(config, (filter, value) => this.onActivateFilter(filter, value)));
        return this;
    }
    addFilters(configs) {
        for (const config of configs) {
            this.addFilter(config);
        }
        return this;
    }
    addIdFilter() {
        this.addFilter({
            name: 'id',
            type: { kind: 'id' },
            label: _('common.id'),
            filterField: 'id',
        });
        return this;
    }
    addDateFilters() {
        this.addFilter({
            name: 'createdAt',
            type: { kind: 'dateRange' },
            label: _('common.created-at'),
            filterField: 'createdAt',
        });
        this.addFilter({
            name: 'updatedAt',
            type: { kind: 'dateRange' },
            label: _('common.updated-at'),
            filterField: 'updatedAt',
        });
        return this;
    }
    addCustomFieldFilters(customFields) {
        for (const config of customFields) {
            const type = config.type;
            if (config.list) {
                continue;
            }
            let filterType;
            switch (type) {
                case 'boolean':
                    filterType = { kind: 'boolean' };
                    break;
                case 'int':
                case 'float':
                    filterType = { kind: 'number' };
                    break;
                case 'datetime':
                    filterType = { kind: 'dateRange' };
                    break;
                case 'string':
                case 'localeString':
                case 'localeText':
                case 'text':
                    filterType = { kind: 'text' };
                    break;
                case 'relation':
                case 'struct':
                    // Cannot filter relations
                    break;
                default:
                    assertNever(type);
            }
            if (filterType) {
                this.addFilter({
                    name: config.name,
                    type: filterType,
                    label: config.label ?? config.name,
                    filterField: config.name,
                });
            }
        }
        return this;
    }
    getFilter(name) {
        return this.#filters.find(f => f.name === name);
    }
    getFilters() {
        return this.#filters;
    }
    removeActiveFilterAtIndex(index) {
        this.#activeFilters.splice(index, 1);
        this.#valueChanges$.next(this.#activeFilters);
    }
    createFilterInput() {
        return this.#activeFilters.reduce((acc, { filter, value }) => {
            const newValue = value != null ? filter.toFilterInput(value) : {};
            const result = extend(true, acc, newValue);
            return result;
        }, {});
    }
    connectToRoute(route) {
        this.valueChanges.pipe(takeUntil(this.destroy$)).subscribe(val => {
            const currentFilters = route.snapshot.queryParamMap.get(this.#filtersQueryParamName);
            if (val.length === 0 && !currentFilters) {
                return;
            }
            this.router.navigate(['./'], {
                queryParams: { [this.#filtersQueryParamName]: this.serialize(), page: 1 },
                relativeTo: route,
                queryParamsHandling: 'merge',
            });
        });
        route.queryParamMap
            .pipe(map(params => params.get(this.#filtersQueryParamName)), distinctUntilChanged(), startWith(route.snapshot.queryParamMap.get(this.#filtersQueryParamName) ?? ''), takeUntil(this.destroy$))
            .subscribe(value => {
            this.#activeFilters = [];
            if (value === '' || value === null) {
                this.#valueChanges$.next(this.#activeFilters);
                return;
            }
            const filterQueryParams = (value ?? '')
                .split(';')
                .map(value => value.split(':'))
                .map(([name, value]) => ({ name, value }));
            for (const { name, value } of filterQueryParams) {
                const filter = this.getFilter(name);
                if (filter) {
                    const val = this.deserializeValue(filter, value);
                    filter.activate(val);
                }
            }
        });
        this.#connectedToRouter = true;
        return this;
    }
    serialize() {
        return this.#activeFilters
            .map((filterWithValue, i) => `${filterWithValue.filter.name}:${this.serializeValue(filterWithValue)}`)
            .join(';');
    }
    serializeValue(filterWithValue) {
        if (filterWithValue.isId()) {
            const val = filterWithValue.value;
            return `${val?.operator},${val?.term}`;
        }
        if (filterWithValue.isText()) {
            const val = filterWithValue.value;
            return `${val?.operator},${val?.term}`;
        }
        else if (filterWithValue.isNumber()) {
            const val = filterWithValue.value;
            return `${val.operator},${val.amount}`;
        }
        else if (filterWithValue.isSelect()) {
            const val = filterWithValue.value;
            return val.join(',');
        }
        else if (filterWithValue.isBoolean()) {
            const val = filterWithValue.value;
            return val ? '1' : '0';
        }
        else if (filterWithValue.isDateRange()) {
            const val = filterWithValue.value;
            if (val.mode === 'relative') {
                return `${val.mode},${val.relativeValue},${val.relativeUnit}`;
            }
            else {
                const start = val.start ? new Date(val.start).getTime() : '';
                const end = val.end ? new Date(val.end).getTime() : '';
                return `${start},${end}`;
            }
        }
        else if (filterWithValue.isCustom()) {
            return filterWithValue.filter.type.serializeValue(filterWithValue.value);
        }
    }
    deserializeValue(filter, value) {
        switch (filter.type.kind) {
            case 'id': {
                const [operator, term] = value.split(',');
                return { operator, term };
            }
            case 'text': {
                const [operator, term] = value.split(',');
                return { operator, term };
            }
            case 'number': {
                const [operator, amount] = value.split(',');
                return { operator, amount: +amount };
            }
            case 'select':
                return value.split(',');
            case 'boolean':
                return value === '1';
            case 'dateRange':
                let mode = 'relative';
                let relativeValue;
                let relativeUnit;
                let start;
                let end;
                if (value.startsWith('relative')) {
                    mode = 'relative';
                    const [_, relativeValueStr, relativeUnitStr] = value.split(',');
                    relativeValue = Number(relativeValueStr);
                    relativeUnit = relativeUnitStr;
                }
                else {
                    mode = 'range';
                    const [startTimestamp, endTimestamp] = value.split(',');
                    start = startTimestamp ? new Date(Number(startTimestamp)).toISOString() : '';
                    end = endTimestamp ? new Date(Number(endTimestamp)).toISOString() : '';
                }
                return { mode, relativeValue, relativeUnit, start, end };
            case 'custom':
                return filter.type.deserializeValue(value);
            default:
                assertNever(filter.type);
        }
    }
    onActivateFilter(filter, value) {
        this.#activeFilters.push(this.createFilterWithValue(filter, value));
        this.#valueChanges$.next(this.#activeFilters);
    }
    createFilterWithValue(filter, value) {
        return new FilterWithValue(filter, value, v => this.#valueChanges$.next(v));
    }
}
