// @vendure/admin-ui/esm2022/core/shared/directives/if-permissions.directive.mjs
import { Directive, Input } from '@angular/core';
import { of } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { IfDirectiveBase } from './if-directive-base';
import * as i0 from "@angular/core";
import * as i1 from "../../providers/permissions/permissions.service";
/**
 * @description
 * Conditionally shows/hides templates based on the current active user having the specified permission.
 * Based on the ngIf source. Also support "else" templates:
 *
 * @example
 * ```html
 * <button *vdrIfPermissions="'DeleteCatalog'; else unauthorized">Delete Product</button>
 * <ng-template #unauthorized>Not allowed!</ng-template>
 * ```
 *
 * The permission can be a single string, or an array. If an array is passed, then _all_ of the permissions
 * must match (logical AND)
 *
 * @docsCategory directives
 */
export class IfPermissionsDirective extends IfDirectiveBase {
    constructor(_viewContainer, templateRef, changeDetectorRef, permissionsService) {
        super(_viewContainer, templateRef, permissions => {
            if (permissions == null) {
                return of(true);
            }
            else if (!permissions) {
                return of(false);
            }
            return this.permissionsService.currentUserPermissions$.pipe(map(() => this.permissionsService.userHasPermissions(permissions)), tap(() => this.changeDetectorRef.markForCheck()));
        });
        this.changeDetectorRef = changeDetectorRef;
        this.permissionsService = permissionsService;
        this.permissionToCheck = ['__initial_value__'];
    }
    /**
     * The permission to check to determine whether to show the template.
     */
    set vdrIfPermissions(permission) {
        this.permissionToCheck =
            (permission && (Array.isArray(permission) ? permission : [permission])) || null;
        this.updateArgs$.next([this.permissionToCheck]);
    }
    /**
     * A template to show if the current user does not have the specified permission.
     */
    set vdrIfPermissionsElse(templateRef) {
        this.setElseTemplate(templateRef);
        this.updateArgs$.next([this.permissionToCheck]);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IfPermissionsDirective, deps: [{ token: i0.ViewContainerRef }, { token: i0.TemplateRef }, { token: i0.ChangeDetectorRef }, { token: i1.PermissionsService }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IfPermissionsDirective, selector: "[vdrIfPermissions]", inputs: { vdrIfPermissions: "vdrIfPermissions", vdrIfPermissionsElse: "vdrIfPermissionsElse" }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IfPermissionsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[vdrIfPermissions]',
                }]
        }], ctorParameters: () => [{ type: i0.ViewContainerRef }, { type: i0.TemplateRef }, { type: i0.ChangeDetectorRef }, { type: i1.PermissionsService }], propDecorators: { vdrIfPermissions: [{
                type: Input
            }], vdrIfPermissionsElse: [{
                type: Input
            }] } });
