// @vendure/admin-ui/esm2022/marketing/marketing.module.mjs
import { AsyncPipe } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule, ROUTES } from '@angular/router';
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { detailComponentWithResolver, GetPromotionDetailDocument, PageService, SharedModule, } from '@vendure/admin-ui/core';
import { PromotionDetailComponent } from './components/promotion-detail/promotion-detail.component';
import { assignPromotionsToChannelBulkAction, deletePromotionsBulkAction, duplicatePromotionsBulkAction, removePromotionsFromChannelBulkAction, } from './components/promotion-list/promotion-list-bulk-actions';
import { PromotionListComponent } from './components/promotion-list/promotion-list.component';
import { createRoutes } from './marketing.routes';
import * as i0 from "@angular/core";
import * as i1 from "@vendure/admin-ui/core";
import * as i2 from "@angular/router";
export class MarketingModule {
    static { this.hasRegisteredTabsAndBulkActions = false; }
    constructor(bulkActionRegistryService, pageService) {
        if (MarketingModule.hasRegisteredTabsAndBulkActions) {
            return;
        }
        bulkActionRegistryService.registerBulkAction(assignPromotionsToChannelBulkAction);
        bulkActionRegistryService.registerBulkAction(duplicatePromotionsBulkAction);
        bulkActionRegistryService.registerBulkAction(removePromotionsFromChannelBulkAction);
        bulkActionRegistryService.registerBulkAction(deletePromotionsBulkAction);
        pageService.registerPageTab({
            priority: 0,
            location: 'promotion-list',
            tab: _('breadcrumb.promotions'),
            route: '',
            component: PromotionListComponent,
        });
        pageService.registerPageTab({
            priority: 0,
            location: 'promotion-detail',
            tab: _('marketing.promotion'),
            route: '',
            component: detailComponentWithResolver({
                component: PromotionDetailComponent,
                query: GetPromotionDetailDocument,
                entityKey: 'promotion',
                getBreadcrumbs: entity => [
                    {
                        label: entity ? entity.name : _('marketing.create-new-promotion'),
                        link: [entity?.id],
                    },
                ],
            }),
        });
        MarketingModule.hasRegisteredTabsAndBulkActions = true;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: MarketingModule, deps: [{ token: i1.BulkActionRegistryService }, { token: i1.PageService }], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.2.4", ngImport: i0, type: MarketingModule, declarations: [PromotionListComponent, PromotionDetailComponent], imports: [SharedModule, i2.RouterModule, SharedModule, AsyncPipe, SharedModule] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: MarketingModule, providers: [
            {
                provide: ROUTES,
                useFactory: (pageService) => createRoutes(pageService),
                multi: true,
                deps: [PageService],
            },
        ], imports: [SharedModule, RouterModule.forChild([]), SharedModule, SharedModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: MarketingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [SharedModule, RouterModule.forChild([]), SharedModule, AsyncPipe, SharedModule],
                    providers: [
                        {
                            provide: ROUTES,
                            useFactory: (pageService) => createRoutes(pageService),
                            multi: true,
                            deps: [PageService],
                        },
                    ],
                    declarations: [PromotionListComponent, PromotionDetailComponent],
                }]
        }], ctorParameters: () => [{ type: i1.BulkActionRegistryService }, { type: i1.PageService }] });
