// @vendure/admin-ui/esm2022/core/shared/components/radio-card/radio-card-fieldset.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
import { Subject } from 'rxjs';
import { throttleTime } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class RadioCardFieldsetComponent {
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this.selectItem = new EventEmitter();
        this.groupName = 'radio-group-' + Math.random().toString(36);
        this.selectedIdChange$ = new Subject();
        this.focussedId = undefined;
        this.idChange$ = new Subject();
    }
    ngOnInit() {
        this.subscription = this.idChange$
            .pipe(throttleTime(200))
            .subscribe(item => this.selectItem.emit(item));
    }
    ngOnChanges(changes) {
        if ('selectedItemId' in changes) {
            this.selectedIdChange$.next(this.selectedItemId);
        }
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    isSelected(item) {
        return this.selectedItemId === this.idFn(item);
    }
    isFocussed(item) {
        return this.focussedId === this.idFn(item);
    }
    selectChanged(item) {
        this.idChange$.next(item);
    }
    setFocussedId(item) {
        this.focussedId = item && this.idFn(item);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: RadioCardFieldsetComponent, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: RadioCardFieldsetComponent, selector: "vdr-radio-card-fieldset", inputs: { selectedItemId: "selectedItemId", idFn: "idFn" }, outputs: { selectItem: "selectItem" }, usesOnChanges: true, ngImport: i0, template: `<fieldset><ng-content></ng-content></fieldset> `, isInline: true, styles: ["fieldset{display:flex;align-items:flex-start;overflow-x:auto}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: RadioCardFieldsetComponent, decorators: [{
            type: Component,
            args: [{ selector: 'vdr-radio-card-fieldset', template: `<fieldset><ng-content></ng-content></fieldset> `, changeDetection: ChangeDetectionStrategy.OnPush, styles: ["fieldset{display:flex;align-items:flex-start;overflow-x:auto}\n"] }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }], propDecorators: { selectedItemId: [{
                type: Input
            }], idFn: [{
                type: Input
            }], selectItem: [{
                type: Output
            }] } });
