// @vendure/admin-ui/esm2022/core/extension/components/angular-route.component.mjs
import { Component, EventEmitter, inject, Input, Output, } from '@angular/core';
import { SharedModule } from '../../shared/shared.module';
import { ROUTE_COMPONENT_OPTIONS, RouteComponent } from './route.component';
import * as i0 from "@angular/core";
/**
 * @description
 * This component is used internally to allow us to dynamically load a component
 * like with `*ngComponentOutlet`, but with the ability to get a reference to the
 * created ComponentRef. This can then be used to delegate lifecycle events like
 * `canDeactivate` to the loaded component.
 */
export class DynamicComponentLoaderComponent {
    constructor(viewContainer) {
        this.viewContainer = viewContainer;
        this.loaded = new EventEmitter();
    }
    ngOnInit() {
        const componentRef = this.viewContainer.createComponent(this.componentType);
        this.loaded.emit(componentRef);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DynamicComponentLoaderComponent, deps: [{ token: i0.ViewContainerRef }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: DynamicComponentLoaderComponent, isStandalone: true, selector: "vdr-dynamic-component-loader", inputs: { componentType: "componentType" }, outputs: { loaded: "loaded" }, ngImport: i0, template: ``, isInline: true, dependencies: [{ kind: "ngmodule", type: SharedModule }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DynamicComponentLoaderComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'vdr-dynamic-component-loader',
                    template: ``,
                    standalone: true,
                    imports: [SharedModule],
                }]
        }], ctorParameters: () => [{ type: i0.ViewContainerRef }], propDecorators: { componentType: [{
                type: Input
            }], loaded: [{
                type: Output
            }] } });
export class AngularRouteComponent {
    constructor() {
        this.component = inject(ROUTE_COMPONENT_OPTIONS).component;
    }
    componentLoaded(componentRef) {
        this.componentRef = componentRef;
    }
    canDeactivate() {
        return this.componentRef?.instance?.canDeactivate?.();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: AngularRouteComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: AngularRouteComponent, isStandalone: true, selector: "vdr-angular-route-component", ngImport: i0, template: `
        <vdr-route-component>
            <vdr-dynamic-component-loader [componentType]="component" (loaded)="componentLoaded($event)" />
        </vdr-route-component>
    `, isInline: true, dependencies: [{ kind: "ngmodule", type: SharedModule }, { kind: "component", type: RouteComponent, selector: "vdr-route-component", inputs: ["locationId", "description"] }, { kind: "component", type: DynamicComponentLoaderComponent, selector: "vdr-dynamic-component-loader", inputs: ["componentType"], outputs: ["loaded"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: AngularRouteComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'vdr-angular-route-component',
                    template: `
        <vdr-route-component>
            <vdr-dynamic-component-loader [componentType]="component" (loaded)="componentLoaded($event)" />
        </vdr-route-component>
    `,
                    standalone: true,
                    imports: [SharedModule, RouteComponent, DynamicComponentLoaderComponent],
                }]
        }] });
