// @vendure/admin-ui/esm2022/core/extension/components/route.component.mjs
import { Component, inject, InjectionToken, Input } from '@angular/core';
import { notNullOrUndefined } from '@vendure/common/lib/shared-utils';
import { combineLatest, Observable, of, switchMap } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { SharedModule } from '../../shared/shared.module';
import { PageMetadataService } from '../providers/page-metadata.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../../providers/page/page.service";
import * as i3 from "@angular/common";
import * as i4 from "../../shared/components/page-header/page-header.component";
import * as i5 from "../../shared/components/page-title/page-title.component";
import * as i6 from "../../shared/components/page-header-description/page-header-description.component";
import * as i7 from "../../shared/components/page-header-tabs/page-header-tabs.component";
import * as i8 from "../../shared/components/page-body/page-body.component";
export const ROUTE_COMPONENT_OPTIONS = new InjectionToken('ROUTE_COMPONENT_OPTIONS');
export class RouteComponent {
    constructor(route, pageService) {
        this.route = route;
        this.pageService = pageService;
        this.headerTabs = [];
        const breadcrumbLabel$ = this.route.data.pipe(switchMap(data => {
            if (data.breadcrumb instanceof Observable) {
                return data.breadcrumb;
            }
            if (typeof data.breadcrumb === 'function') {
                return data.breadcrumb(data);
            }
            return of(undefined);
        }), filter(notNullOrUndefined), map(breadcrumb => {
            if (typeof breadcrumb === 'string') {
                return breadcrumb;
            }
            if (Array.isArray(breadcrumb)) {
                return breadcrumb[breadcrumb.length - 1].label;
            }
            return breadcrumb.label;
        }));
        this.title$ = combineLatest([inject(ROUTE_COMPONENT_OPTIONS).title$, breadcrumbLabel$]).pipe(map(([title, breadcrumbLabel]) => title ?? breadcrumbLabel));
        this.locationId = this.route.snapshot.data.locationId;
        this.description = this.route.snapshot.data.description;
        this.headerTabs = this.pageService.getPageTabs(this.locationId).map(tab => ({
            id: tab.tab,
            label: tab.tab,
            icon: tab.tabIcon,
            route: tab.route ? [tab.route] : ['./'],
        }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: RouteComponent, deps: [{ token: i1.ActivatedRoute }, { token: i2.PageService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: RouteComponent, isStandalone: true, selector: "vdr-route-component", inputs: { locationId: "locationId", description: "description" }, providers: [PageMetadataService], ngImport: i0, template: `
        <vdr-page-header>
            <vdr-page-title *ngIf="title$ | async as title" [title]="title"></vdr-page-title>
            <vdr-page-header-description *ngIf="description">{{ description }}</vdr-page-header-description>
            <vdr-page-header-tabs *ngIf="headerTabs.length > 1" [tabs]="headerTabs"></vdr-page-header-tabs>
        </vdr-page-header>
        <vdr-page-body><ng-content /></vdr-page-body>
    `, isInline: true, dependencies: [{ kind: "ngmodule", type: SharedModule }, { kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i4.PageHeaderComponent, selector: "vdr-page-header" }, { kind: "component", type: i5.PageTitleComponent, selector: "vdr-page-title", inputs: ["title"] }, { kind: "component", type: i6.PageHeaderDescriptionComponent, selector: "vdr-page-header-description" }, { kind: "component", type: i7.PageHeaderTabsComponent, selector: "vdr-page-header-tabs", inputs: ["tabs"] }, { kind: "component", type: i8.PageBodyComponent, selector: "vdr-page-body" }, { kind: "pipe", type: i3.AsyncPipe, name: "async" }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: RouteComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'vdr-route-component',
                    template: `
        <vdr-page-header>
            <vdr-page-title *ngIf="title$ | async as title" [title]="title"></vdr-page-title>
            <vdr-page-header-description *ngIf="description">{{ description }}</vdr-page-header-description>
            <vdr-page-header-tabs *ngIf="headerTabs.length > 1" [tabs]="headerTabs"></vdr-page-header-tabs>
        </vdr-page-header>
        <vdr-page-body><ng-content /></vdr-page-body>
    `,
                    standalone: true,
                    imports: [SharedModule],
                    providers: [PageMetadataService],
                }]
        }], ctorParameters: () => [{ type: i1.ActivatedRoute }, { type: i2.PageService }], propDecorators: { locationId: [{
                type: Input
            }], description: [{
                type: Input
            }] } });
