// @vendure/admin-ui/esm2022/core/shared/components/dropdown/dropdown-item.directive.mjs
import { Directive, HostListener, Inject } from '@angular/core';
import { DropdownComponent } from './dropdown.component';
import * as i0 from "@angular/core";
export class DropdownItemDirective {
    constructor(dropdown) {
        this.dropdown = dropdown;
    }
    async onDropdownItemClick() {
        (await this.dropdown).onClick();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DropdownItemDirective, deps: [{ token: DropdownComponent }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: DropdownItemDirective, selector: "[vdrDropdownItem]", host: { listeners: { "click": "onDropdownItemClick($event)" }, properties: { "class.dropdown-item": "true" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DropdownItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[vdrDropdownItem]',
                    // eslint-disable-next-line
                    host: { '[class.dropdown-item]': 'true' },
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DropdownComponent]
                }] }], propDecorators: { onDropdownItemClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
