// @vendure/admin-ui/esm2022/core/providers/permissions/permissions.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * @description
 * This service is used internally to power components & logic that are dependent on knowing the
 * current user's permissions in the currently-active channel.
 */
export class PermissionsService {
    constructor() {
        this.currentUserPermissions = [];
        this._currentUserPermissions$ = new BehaviorSubject([]);
        this.currentUserPermissions$ = this._currentUserPermissions$.asObservable();
    }
    /**
     * @description
     * This is called whenever:
     * - the user logs in
     * - the active channel changes
     *
     * Since active user validation occurs as part of the main auth guard, we can be assured
     * that if the user is logged in, then this method will be called with the user's permissions
     * before any other components are rendered lower down in the component tree.
     */
    setCurrentUserPermissions(permissions) {
        this.currentUserPermissions = permissions;
        this._currentUserPermissions$.next(permissions);
    }
    userHasPermissions(requiredPermissions) {
        for (const perm of requiredPermissions) {
            if (this.currentUserPermissions.includes(perm)) {
                return true;
            }
        }
        return false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: PermissionsService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: PermissionsService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: PermissionsService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
