// @vendure/admin-ui/esm2022/customer/customer.routes.mjs
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { detailBreadcrumb, PageComponent } from '@vendure/admin-ui/core';
export const createRoutes = (pageService) => [
    {
        path: 'customers',
        component: PageComponent,
        data: {
            locationId: 'customer-list',
            breadcrumb: _('breadcrumb.customers'),
        },
        children: pageService.getPageTabRoutes('customer-list'),
    },
    {
        path: 'customers/:id',
        component: PageComponent,
        data: {
            locationId: 'customer-detail',
            breadcrumb: { label: _('breadcrumb.customers'), link: ['../', 'customers'] },
        },
        children: pageService.getPageTabRoutes('customer-detail'),
    },
    {
        path: 'groups',
        component: PageComponent,
        data: {
            locationId: 'customer-group-list',
            breadcrumb: _('breadcrumb.customer-groups'),
        },
        children: pageService.getPageTabRoutes('customer-group-list'),
    },
    {
        path: 'groups/:id',
        component: PageComponent,
        data: {
            locationId: 'customer-group-detail',
            breadcrumb: { label: _('breadcrumb.customer-groups'), link: ['../', 'groups'] },
        },
        children: pageService.getPageTabRoutes('customer-group-detail'),
    },
];
export function customerBreadcrumb(data, params) {
    return detailBreadcrumb({
        entity: data.entity,
        id: params.id,
        breadcrumbKey: 'breadcrumb.customers',
        getName: customer => `${customer.firstName} ${customer.lastName}`,
        route: 'customers',
    });
}
