// @vendure/admin-ui/esm2022/customer/customer.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule, ROUTES } from '@angular/router';
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { CustomerDetailQueryDocument, detailComponentWithResolver, GetCustomerGroupDetailDocument, PageService, SharedModule, SortOrder, } from '@vendure/admin-ui/core';
import { AddCustomerToGroupDialogComponent } from './components/add-customer-to-group-dialog/add-customer-to-group-dialog.component';
import { AddressCardComponent } from './components/address-card/address-card.component';
import { AddressDetailDialogComponent } from './components/address-detail-dialog/address-detail-dialog.component';
import { CustomerDetailComponent } from './components/customer-detail/customer-detail.component';
import { CustomerGroupDetailDialogComponent } from './components/customer-group-detail-dialog/customer-group-detail-dialog.component';
import { deleteCustomerGroupsBulkAction } from './components/customer-group-list/customer-group-list-bulk-actions';
import { CustomerGroupListComponent } from './components/customer-group-list/customer-group-list.component';
import { removeCustomerGroupMembersBulkAction } from './components/customer-group-member-list/customer-group-member-list-bulk-actions';
import { CustomerGroupMemberListComponent } from './components/customer-group-member-list/customer-group-member-list.component';
import { CustomerHistoryEntryHostComponent } from './components/customer-history/customer-history-entry-host.component';
import { CustomerHistoryComponent } from './components/customer-history/customer-history.component';
import { deleteCustomersBulkAction } from './components/customer-list/customer-list-bulk-actions';
import { CustomerListComponent } from './components/customer-list/customer-list.component';
import { CustomerStatusLabelComponent } from './components/customer-status-label/customer-status-label.component';
import { SelectCustomerGroupDialogComponent } from './components/select-customer-group-dialog/select-customer-group-dialog.component';
import { createRoutes } from './customer.routes';
import { CustomerGroupDetailComponent } from './components/customer-group-detail/customer-group-detail.component';
import * as i0 from "@angular/core";
import * as i1 from "@vendure/admin-ui/core";
import * as i2 from "@angular/router";
export class CustomerModule {
    static { this.hasRegisteredTabsAndBulkActions = false; }
    constructor(bulkActionRegistryService, pageService) {
        if (CustomerModule.hasRegisteredTabsAndBulkActions) {
            return;
        }
        bulkActionRegistryService.registerBulkAction(deleteCustomersBulkAction);
        bulkActionRegistryService.registerBulkAction(deleteCustomerGroupsBulkAction);
        bulkActionRegistryService.registerBulkAction(removeCustomerGroupMembersBulkAction);
        pageService.registerPageTab({
            priority: 0,
            location: 'customer-list',
            tab: _('customer.customers'),
            route: '',
            component: CustomerListComponent,
        });
        pageService.registerPageTab({
            priority: 0,
            location: 'customer-detail',
            tab: _('customer.customer'),
            route: '',
            component: detailComponentWithResolver({
                component: CustomerDetailComponent,
                query: CustomerDetailQueryDocument,
                entityKey: 'customer',
                variables: {
                    orderListOptions: {
                        sort: {
                            orderPlacedAt: SortOrder.DESC,
                        },
                    },
                },
                getBreadcrumbs: entity => [
                    {
                        label: entity
                            ? `${entity?.firstName} ${entity?.lastName}`
                            : _('customer.create-new-customer'),
                        link: [entity?.id],
                    },
                ],
            }),
        });
        pageService.registerPageTab({
            priority: 0,
            location: 'customer-group-list',
            tab: _('customer.customer-groups'),
            route: '',
            component: CustomerGroupListComponent,
        });
        pageService.registerPageTab({
            priority: 0,
            location: 'customer-group-detail',
            tab: _('customer.customer-group'),
            route: '',
            component: detailComponentWithResolver({
                component: CustomerGroupDetailComponent,
                query: GetCustomerGroupDetailDocument,
                entityKey: 'customerGroup',
                getBreadcrumbs: entity => [
                    {
                        label: entity ? entity.name : _('customer.create-new-customer-group'),
                        link: [entity?.id],
                    },
                ],
            }),
        });
        CustomerModule.hasRegisteredTabsAndBulkActions = true;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: CustomerModule, deps: [{ token: i1.BulkActionRegistryService }, { token: i1.PageService }], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.2.4", ngImport: i0, type: CustomerModule, declarations: [CustomerListComponent,
            CustomerDetailComponent,
            CustomerStatusLabelComponent,
            AddressCardComponent,
            CustomerGroupListComponent,
            CustomerGroupDetailDialogComponent,
            AddCustomerToGroupDialogComponent,
            CustomerGroupMemberListComponent,
            SelectCustomerGroupDialogComponent,
            CustomerHistoryComponent,
            AddressDetailDialogComponent,
            CustomerHistoryEntryHostComponent,
            CustomerGroupDetailComponent], imports: [SharedModule, i2.RouterModule], exports: [AddressCardComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: CustomerModule, providers: [
            {
                provide: ROUTES,
                useFactory: (pageService) => createRoutes(pageService),
                multi: true,
                deps: [PageService],
            },
        ], imports: [SharedModule, RouterModule.forChild([])] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: CustomerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [SharedModule, RouterModule.forChild([])],
                    providers: [
                        {
                            provide: ROUTES,
                            useFactory: (pageService) => createRoutes(pageService),
                            multi: true,
                            deps: [PageService],
                        },
                    ],
                    declarations: [
                        CustomerListComponent,
                        CustomerDetailComponent,
                        CustomerStatusLabelComponent,
                        AddressCardComponent,
                        CustomerGroupListComponent,
                        CustomerGroupDetailDialogComponent,
                        AddCustomerToGroupDialogComponent,
                        CustomerGroupMemberListComponent,
                        SelectCustomerGroupDialogComponent,
                        CustomerHistoryComponent,
                        AddressDetailDialogComponent,
                        CustomerHistoryEntryHostComponent,
                        CustomerGroupDetailComponent,
                    ],
                    exports: [AddressCardComponent],
                }]
        }], ctorParameters: () => [{ type: i1.BulkActionRegistryService }, { type: i1.PageService }] });
