// @vendure/admin-ui/esm2022/core/shared/components/rich-text-editor/rich-text-editor.component.mjs
import { ChangeDetectionStrategy, Component, HostBinding, Input, ViewChild, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ContextMenuService } from './prosemirror/context-menu/context-menu.service';
import { ProsemirrorService } from './prosemirror/prosemirror.service';
import * as i0 from "@angular/core";
import * as i1 from "./prosemirror/prosemirror.service";
import * as i2 from "./prosemirror/context-menu/context-menu.service";
import * as i3 from "@clr/angular";
import * as i4 from "@angular/common";
import * as i5 from "./prosemirror/context-menu/context-menu.component";
/**
 * @description
 * A rich text (HTML) editor based on Prosemirror (https://prosemirror.net/)
 *
 * @example
 * ```HTML
 * <vdr-rich-text-editor
 *     [(ngModel)]="description"
 *     label="Description"
 * ></vdr-rich-text-editor>
 * ```
 *
 * @docsCategory components
 */
export class RichTextEditorComponent {
    set readonly(value) {
        this._readonly = !!value;
        this.prosemirrorService.setEnabled(!this._readonly);
    }
    constructor(changeDetector, prosemirrorService, viewContainerRef, contextMenuService) {
        this.changeDetector = changeDetector;
        this.prosemirrorService = prosemirrorService;
        this.viewContainerRef = viewContainerRef;
        this.contextMenuService = contextMenuService;
        this._readonly = false;
    }
    get menuElement() {
        return this.viewContainerRef.element.nativeElement.querySelector('.ProseMirror-menubar');
    }
    ngAfterViewInit() {
        this.prosemirrorService.createEditorView({
            element: this.editorEl.nativeElement,
            onTextInput: content => {
                this.onChange(content);
                this.changeDetector.markForCheck();
            },
            isReadOnly: () => !this._readonly,
        });
        if (this.value) {
            this.prosemirrorService.update(this.value);
        }
    }
    ngOnDestroy() {
        this.prosemirrorService.destroy();
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouch = fn;
    }
    setDisabledState(isDisabled) {
        this.prosemirrorService.setEnabled(!isDisabled);
    }
    writeValue(value) {
        if (value !== this.value) {
            this.value = value;
            if (this.prosemirrorService) {
                this.prosemirrorService.update(value);
            }
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: RichTextEditorComponent, deps: [{ token: i0.ChangeDetectorRef }, { token: i1.ProsemirrorService }, { token: i0.ViewContainerRef }, { token: i2.ContextMenuService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: RichTextEditorComponent, selector: "vdr-rich-text-editor", inputs: { label: "label", readonly: "readonly" }, host: { properties: { "class.readonly": "this._readonly" } }, providers: [
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: RichTextEditorComponent,
                multi: true,
            },
            ProsemirrorService,
            ContextMenuService,
        ], viewQueries: [{ propertyName: "editorEl", first: true, predicate: ["editor"], descendants: true, static: true }], ngImport: i0, template: "<label *ngIf=\"label\" class=\"rich-text-label\">{{ label }}</label>\r\n<div #editor></div>\r\n<vdr-context-menu [editorMenuElement]=\"menuElement\"></vdr-context-menu>\r\n", styles: [":host{display:block;margin-bottom:.5rem}:host.readonly ::ng-deep .ProseMirror-menubar{display:none}.context-menu{position:fixed}\n"], dependencies: [{ kind: "directive", type: i3.ClrLabel, selector: "label", inputs: ["for"] }, { kind: "directive", type: i4.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i5.ContextMenuComponent, selector: "vdr-context-menu", inputs: ["editorMenuElement"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: RichTextEditorComponent, decorators: [{
            type: Component,
            args: [{ selector: 'vdr-rich-text-editor', changeDetection: ChangeDetectionStrategy.OnPush, providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: RichTextEditorComponent,
                            multi: true,
                        },
                        ProsemirrorService,
                        ContextMenuService,
                    ], template: "<label *ngIf=\"label\" class=\"rich-text-label\">{{ label }}</label>\r\n<div #editor></div>\r\n<vdr-context-menu [editorMenuElement]=\"menuElement\"></vdr-context-menu>\r\n", styles: [":host{display:block;margin-bottom:.5rem}:host.readonly ::ng-deep .ProseMirror-menubar{display:none}.context-menu{position:fixed}\n"] }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }, { type: i1.ProsemirrorService }, { type: i0.ViewContainerRef }, { type: i2.ContextMenuService }], propDecorators: { label: [{
                type: Input
            }], readonly: [{
                type: Input
            }], _readonly: [{
                type: HostBinding,
                args: ['class.readonly']
            }], editorEl: [{
                type: ViewChild,
                args: ['editor', { static: true }]
            }] } });
