// @vendure/admin-ui/esm2022/core/shared/components/data-table-filter-presets/filter-preset.service.mjs
import { moveItemInArray } from '@angular/cdk/drag-drop';
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../../../providers/data-table/data-table-config.service";
export class FilterPresetService {
    constructor(dataTableConfigService) {
        this.dataTableConfigService = dataTableConfigService;
        this._presetChanges = new Subject();
        this.presetChanges$ = this._presetChanges.asObservable();
    }
    getFilterPresets(dataTableId) {
        const dataTableConfig = this.dataTableConfigService.getConfig(dataTableId);
        return dataTableConfig.filterPresets ?? [];
    }
    saveFilterPreset(config) {
        const dataTableConfig = this.dataTableConfigService.getConfig(config.dataTableId);
        const filterPresets = dataTableConfig.filterPresets ?? [];
        const existingName = filterPresets.find(p => p.name === config.name);
        if (existingName) {
            existingName.value = config.value;
        }
        else {
            filterPresets.push({
                name: config.name,
                value: config.value,
            });
        }
        dataTableConfig.filterPresets = filterPresets;
        this.dataTableConfigService.setConfig(config.dataTableId, dataTableConfig);
        this._presetChanges.next(filterPresets);
    }
    deleteFilterPreset(config) {
        const dataTableConfig = this.dataTableConfigService.getConfig(config.dataTableId);
        dataTableConfig.filterPresets = dataTableConfig.filterPresets.filter(p => p.name !== config.name);
        this.dataTableConfigService.setConfig(config.dataTableId, dataTableConfig);
        this._presetChanges.next(dataTableConfig.filterPresets);
    }
    reorderPresets(dataTableId, fromIndex, toIndex) {
        const presets = this.getFilterPresets(dataTableId);
        moveItemInArray(presets, fromIndex, toIndex);
        const dataTableConfig = this.dataTableConfigService.getConfig(dataTableId);
        dataTableConfig.filterPresets = presets;
        this.dataTableConfigService.setConfig(dataTableId, dataTableConfig);
        this._presetChanges.next(presets);
    }
    renameFilterPreset(config) {
        const dataTableConfig = this.dataTableConfigService.getConfig(config.dataTableId);
        const filterPresets = dataTableConfig.filterPresets ?? [];
        const existingName = filterPresets.find(p => p.name === config.oldName);
        if (existingName) {
            existingName.name = config.newName;
            dataTableConfig.filterPresets = filterPresets;
            this.dataTableConfigService.setConfig(config.dataTableId, dataTableConfig);
            this._presetChanges.next(filterPresets);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: FilterPresetService, deps: [{ token: i1.DataTableConfigService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: FilterPresetService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: FilterPresetService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [{ type: i1.DataTableConfigService }] });
