// @vendure/admin-ui/esm2022/core/shared/components/duplicate-entity-dialog/duplicate-entity-dialog.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FormControl } from '@angular/forms';
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { assertNever } from '@vendure/common/lib/shared-utils';
import { lastValueFrom } from 'rxjs';
import { tap } from 'rxjs/operators';
import { DuplicateEntityDocument, GetEntityDuplicatorsDocument, } from '../../../common/generated-types';
import { configurableDefinitionToInstance, toConfigurableOperationInput, } from '../../../common/utilities/configurable-operation-utils';
import * as i0 from "@angular/core";
import * as i1 from "../../../data/providers/data.service";
import * as i2 from "../../../providers/notification/notification.service";
import * as i3 from "@angular/common";
import * as i4 from "@angular/forms";
import * as i5 from "../configurable-input/configurable-input.component";
import * as i6 from "../form-field/form-field-control.directive";
import * as i7 from "../modal-dialog/dialog-buttons.directive";
import * as i8 from "../modal-dialog/dialog-title.directive";
import * as i9 from "@ngx-translate/core";
export class DuplicateEntityDialogComponent {
    constructor(dataService, notificationService, changeDetectorRef) {
        this.dataService = dataService;
        this.notificationService = notificationService;
        this.changeDetectorRef = changeDetectorRef;
        this.formGroup = new FormControl({
            code: '',
            args: [],
        });
    }
    ngOnInit() {
        this.entityDuplicators$ = this.dataService
            .query(GetEntityDuplicatorsDocument)
            .mapSingle(data => data.entityDuplicators.filter(d => d.forEntities.includes(this.entityName)))
            .pipe(tap(duplicators => {
            if (0 < duplicators.length) {
                this.setSelectedDuplicator(duplicators[0]);
            }
        }));
    }
    setSelectedDuplicator(duplicator) {
        this.selectedDuplicator = duplicator;
        this.duplicatorInstance = configurableDefinitionToInstance(this.selectedDuplicator);
        this.formGroup.patchValue(this.duplicatorInstance);
        this.changeDetectorRef.markForCheck();
    }
    async duplicate() {
        const selectedDuplicator = this.selectedDuplicator;
        const formValue = this.formGroup.value;
        if (!selectedDuplicator || !formValue) {
            return;
        }
        const duplicatorInput = toConfigurableOperationInput(this.duplicatorInstance, formValue);
        const succeeded = [];
        const failed = [];
        for (const entity of this.entities) {
            const { duplicateEntity } = await lastValueFrom(this.dataService.mutate(DuplicateEntityDocument, {
                input: {
                    entityId: entity.id,
                    entityName: this.entityName,
                    duplicatorInput,
                },
            }));
            switch (duplicateEntity.__typename) {
                case 'DuplicateEntitySuccess':
                    succeeded.push(this.getEntityName(entity));
                    break;
                case 'DuplicateEntityError':
                    failed.push({
                        name: this.getEntityName(entity),
                        message: duplicateEntity.duplicationError,
                    });
                    break;
                case undefined:
                    break;
                default:
                    assertNever(duplicateEntity);
            }
        }
        if (0 < succeeded.length) {
            this.notificationService.success(_('common.notify-duplicate-success'), {
                count: succeeded.length,
                names: succeeded.join(', '),
            });
        }
        if (0 < failed.length) {
            const failedCount = failed.length;
            const maxNotices = 5;
            const excess = failedCount - maxNotices;
            for (let i = 0; i < Math.min(failedCount, maxNotices); i++) {
                const failedItem = failed[i];
                this.notificationService.error(_('common.notify-duplicate-error'), {
                    name: failedItem.name,
                    error: failedItem.message,
                });
            }
            if (excess > 0) {
                this.notificationService.error(_('common.notify-duplicate-error-excess'), { count: excess });
            }
        }
        this.resolveWith(true);
    }
    cancel() {
        this.resolveWith();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DuplicateEntityDialogComponent, deps: [{ token: i1.DataService }, { token: i2.NotificationService }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: DuplicateEntityDialogComponent, selector: "vdr-duplicate-entity-dialog", ngImport: i0, template: "<ng-template vdrDialogTitle>\r\n    {{ title || 'common.duplicate-entity' | translate }}\r\n</ng-template>\r\n<div class=\"\">\r\n    <ul>\r\n        <li *ngFor=\"let entity of entities\">\r\n            {{ getEntityName(entity) }}\r\n        </li>\r\n    </ul>\r\n    <ng-container *ngIf=\"entityDuplicators$ | async as duplicators\">\r\n        <select\r\n            name=\"\"\r\n            id=\"\"\r\n            *ngIf=\"1 < duplicators.length\"\r\n            class=\"mt-4\"\r\n            [ngModel]=\"selectedDuplicator\"\r\n            (ngModelChange)=\"setSelectedDuplicator($event)\"\r\n        >\r\n            <option *ngFor=\"let duplicator of duplicators\" [ngValue]=\"duplicator\">\r\n                {{ duplicator.description }} ({{ duplicator.code }})\r\n            </option>\r\n        </select>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"selectedDuplicator as duplicator\">\r\n        <vdr-configurable-input\r\n            *ngIf=\"duplicator.args.length > 0\"\r\n            [hideDescription]=\"true\"\r\n            [operation]=\"duplicatorInstance\"\r\n            [operationDefinition]=\"duplicator\"\r\n            [removable]=\"false\"\r\n            [formControl]=\"formGroup\"\r\n        ></vdr-configurable-input>\r\n    </ng-container>\r\n</div>\r\n\r\n<ng-template vdrDialogButtons>\r\n    <button type=\"button\" class=\"btn\" (click)=\"cancel()\">{{ 'common.cancel' | translate }}</button>\r\n    <button type=\"submit\" (click)=\"duplicate()\" [disabled]=\"!selectedDuplicator\" class=\"btn btn-primary\">\r\n        {{ 'common.duplicate' | translate }}\r\n    </button>\r\n</ng-template>\r\n", styles: [""], dependencies: [{ kind: "directive", type: i3.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i4.NgSelectOption, selector: "option", inputs: ["ngValue", "value"] }, { kind: "directive", type: i4.ɵNgSelectMultipleOption, selector: "option", inputs: ["ngValue", "value"] }, { kind: "directive", type: i4.SelectControlValueAccessor, selector: "select:not([multiple])[formControlName],select:not([multiple])[formControl],select:not([multiple])[ngModel]", inputs: ["compareWith"] }, { kind: "directive", type: i4.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i4.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { kind: "directive", type: i4.FormControlDirective, selector: "[formControl]", inputs: ["formControl", "disabled", "ngModel"], outputs: ["ngModelChange"], exportAs: ["ngForm"] }, { kind: "component", type: i5.ConfigurableInputComponent, selector: "vdr-configurable-input", inputs: ["operation", "operationDefinition", "readonly", "removable", "position", "hideDescription"], outputs: ["remove"] }, { kind: "directive", type: i6.FormFieldControlDirective, selector: "input, textarea, select, vdr-currency-input" }, { kind: "directive", type: i7.DialogButtonsDirective, selector: "[vdrDialogButtons]" }, { kind: "directive", type: i8.DialogTitleDirective, selector: "[vdrDialogTitle]" }, { kind: "pipe", type: i3.AsyncPipe, name: "async" }, { kind: "pipe", type: i9.TranslatePipe, name: "translate" }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: DuplicateEntityDialogComponent, decorators: [{
            type: Component,
            args: [{ selector: 'vdr-duplicate-entity-dialog', changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-template vdrDialogTitle>\r\n    {{ title || 'common.duplicate-entity' | translate }}\r\n</ng-template>\r\n<div class=\"\">\r\n    <ul>\r\n        <li *ngFor=\"let entity of entities\">\r\n            {{ getEntityName(entity) }}\r\n        </li>\r\n    </ul>\r\n    <ng-container *ngIf=\"entityDuplicators$ | async as duplicators\">\r\n        <select\r\n            name=\"\"\r\n            id=\"\"\r\n            *ngIf=\"1 < duplicators.length\"\r\n            class=\"mt-4\"\r\n            [ngModel]=\"selectedDuplicator\"\r\n            (ngModelChange)=\"setSelectedDuplicator($event)\"\r\n        >\r\n            <option *ngFor=\"let duplicator of duplicators\" [ngValue]=\"duplicator\">\r\n                {{ duplicator.description }} ({{ duplicator.code }})\r\n            </option>\r\n        </select>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"selectedDuplicator as duplicator\">\r\n        <vdr-configurable-input\r\n            *ngIf=\"duplicator.args.length > 0\"\r\n            [hideDescription]=\"true\"\r\n            [operation]=\"duplicatorInstance\"\r\n            [operationDefinition]=\"duplicator\"\r\n            [removable]=\"false\"\r\n            [formControl]=\"formGroup\"\r\n        ></vdr-configurable-input>\r\n    </ng-container>\r\n</div>\r\n\r\n<ng-template vdrDialogButtons>\r\n    <button type=\"button\" class=\"btn\" (click)=\"cancel()\">{{ 'common.cancel' | translate }}</button>\r\n    <button type=\"submit\" (click)=\"duplicate()\" [disabled]=\"!selectedDuplicator\" class=\"btn btn-primary\">\r\n        {{ 'common.duplicate' | translate }}\r\n    </button>\r\n</ng-template>\r\n" }]
        }], ctorParameters: () => [{ type: i1.DataService }, { type: i2.NotificationService }, { type: i0.ChangeDetectorRef }] });
