// @vendure/admin-ui/esm2022/react/react-components/RichTextEditor.mjs
import React, { forwardRef, useEffect, useState, } from 'react';
import { useRichTextEditor } from '../react-hooks/use-rich-text-editor';
/**
 * @description
 * A rich text editor component which uses ProseMirror (rich text editor) under the hood.
 *
 * @example
 * ```ts
 * import { RichTextEditor } from '@vendure/admin-ui/react';
 * import React from 'react';
 *
 * export function MyComponent() {
 *   const onSubmit = async (e: React.FormEvent) => {
 *     e.preventDefault();
 *     const form = new FormData(e.target as HTMLFormElement);
 *     const content = form.get("content");
 *     console.log(content);
 *   };
 *
 *   return (
 *     <form className="w-full" onSubmit={onSubmit}>
 *       <RichTextEditor
 *         name="content"
 *         readOnly={false}
 *         onMount={(e) => console.log("Mounted", e)}
 *       />
 *       <button type="submit" className="btn btn-primary">
 *         Submit
 *       </button>
 *     </form>
 *   );
 * }
 * ```
 *
 * @docsCategory react-components
 */
export const RichTextEditor = forwardRef((props, ref) => {
    const [data, setData] = useState('');
    const { readOnly, label, ...rest } = props;
    const { ref: _ref, editor } = useRichTextEditor({
        attributes: props.attributes,
        isReadOnly: () => readOnly || false,
        onTextInput: text => {
            setData(text);
            if (props.onChange) {
                props.onChange({
                    target: { value: text },
                });
            }
            if (ref && 'current' in ref && ref.current) {
                ref.current.value = text;
                const event = new Event('input', {
                    bubbles: true,
                    cancelable: true,
                });
                ref.current.dispatchEvent(event);
            }
        },
    });
    useEffect(() => {
        if (props.onMount && editor) {
            props.onMount(editor);
        }
        if (typeof props.defaultValue === 'string') {
            editor.update(props.defaultValue);
        }
    }, []);
    return (React.createElement(React.Fragment, null,
        React.createElement("div", { ref: _ref, ...rest }, label && React.createElement("label", { className: "rich-text-label" }, label)),
        React.createElement("input", { type: "hidden", value: data, ref: ref })));
});
RichTextEditor.displayName = 'RichTextEditor';
