// @vendure/admin-ui/esm2022/core/shared/components/page-title/page-title.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { titleSetter } from '../../../common/title-setter';
import * as i0 from "@angular/core";
import * as i1 from "../../../providers/breadcrumb/breadcrumb.service";
import * as i2 from "@angular/common";
import * as i3 from "@ngx-translate/core";
export class PageTitleComponent {
    constructor(breadcrumbService) {
        this.breadcrumbService = breadcrumbService;
        this.title = '';
        this.titleChange$ = new BehaviorSubject(undefined);
        this.setTitle = titleSetter();
    }
    ngOnInit() {
        this.title$ = combineLatest(this.titleChange$, this.breadcrumbService.breadcrumbs$).pipe(map(([title, breadcrumbs]) => {
            if (title) {
                return title;
            }
            else {
                return breadcrumbs[breadcrumbs.length - 1].label;
            }
        }), tap(title => this.setTitle(title)));
    }
    ngOnChanges(changes) {
        if (changes.title) {
            this.titleChange$.next(changes.title.currentValue);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: PageTitleComponent, deps: [{ token: i1.BreadcrumbService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: PageTitleComponent, selector: "vdr-page-title", inputs: { title: "title" }, usesOnChanges: true, ngImport: i0, template: "<div class=\"page-title\">\r\n    <h1>{{ (title$ | async) ?? '' | translate }}</h1>\r\n    <div class=\"title-actions\"><ng-content></ng-content></div>\r\n</div>\r\n", styles: [":host{display:block}.page-title{display:flex;gap:calc(var(--space-unit) * 2);margin-bottom:calc(var(--space-unit) * 2)}.page-title h1{margin-top:0;color:var(--color-weight-900);font-size:var(--cds-global-typography-headline-font-size);font-weight:600;line-height:48px}@media screen and (max-width: 768px){.page-title h1{font-size:var(--font-size-xl)}}@media screen and (max-width: 992px){.page-title h1{font-size:24px}}.title-actions{display:flex;align-items:center}\n"], dependencies: [{ kind: "pipe", type: i2.AsyncPipe, name: "async" }, { kind: "pipe", type: i3.TranslatePipe, name: "translate" }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: PageTitleComponent, decorators: [{
            type: Component,
            args: [{ selector: 'vdr-page-title', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"page-title\">\r\n    <h1>{{ (title$ | async) ?? '' | translate }}</h1>\r\n    <div class=\"title-actions\"><ng-content></ng-content></div>\r\n</div>\r\n", styles: [":host{display:block}.page-title{display:flex;gap:calc(var(--space-unit) * 2);margin-bottom:calc(var(--space-unit) * 2)}.page-title h1{margin-top:0;color:var(--color-weight-900);font-size:var(--cds-global-typography-headline-font-size);font-weight:600;line-height:48px}@media screen and (max-width: 768px){.page-title h1{font-size:var(--font-size-xl)}}@media screen and (max-width: 992px){.page-title h1{font-size:24px}}.title-actions{display:flex;align-items:center}\n"] }]
        }], ctorParameters: () => [{ type: i1.BreadcrumbService }], propDecorators: { title: [{
                type: Input
            }] } });
