// @vendure/admin-ui/esm2022/core/shared/components/rich-text-editor/prosemirror/menu/menu-common.mjs
import { notNullOrUndefined } from '@vendure/common/lib/shared-utils';
export function markActive(state, type) {
    const { from, $from, to, empty } = state.selection;
    if (empty) {
        return type.isInSet(state.storedMarks || $from.marks());
    }
    else {
        return state.doc.rangeHasMark(from, to, type);
    }
}
export function canInsert(state, nodeType) {
    const $from = state.selection.$from;
    for (let d = $from.depth; d >= 0; d--) {
        const index = $from.index(d);
        if ($from.node(d).canReplaceWith(index, index, nodeType)) {
            return true;
        }
    }
    return false;
}
export function renderClarityIcon(options) {
    return (view) => {
        const icon = document.createElement('clr-icon');
        icon.setAttribute('shape', options.shape);
        icon.setAttribute('size', (options.size ?? IconSize.Small).toString());
        const labelEl = document.createElement('span');
        labelEl.textContent = options.label ?? '';
        return wrapInMenuItemWithIcon(icon, options.label ? labelEl : undefined);
    };
}
export function wrapInMenuItemWithIcon(...elements) {
    const wrapperEl = document.createElement('span');
    wrapperEl.classList.add('menu-item-with-icon');
    wrapperEl.append(...elements.filter(notNullOrUndefined));
    return wrapperEl;
}
export const IconSize = {
    Large: 22,
    Small: 16,
};
