// @vendure/admin-ui/esm2022/core/providers/channel/channel.service.mjs
import { Injectable } from '@angular/core';
import { DEFAULT_CHANNEL_CODE } from '@vendure/common/lib/shared-constants';
import { map, shareReplay, tap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "../../data/providers/data.service";
import * as i2 from "../local-storage/local-storage.service";
import * as i3 from "../permissions/permissions.service";
export class ChannelService {
    constructor(dataService, localStorageService, permissionsService) {
        this.dataService = dataService;
        this.localStorageService = localStorageService;
        this.permissionsService = permissionsService;
        this.defaultChannelIsActive$ = this.dataService.client
            .userStatus()
            .mapStream(({ userStatus }) => {
            const activeChannel = userStatus.channels.find(c => c.id === userStatus.activeChannelId);
            return activeChannel ? activeChannel.code === DEFAULT_CHANNEL_CODE : false;
        })
            .pipe(shareReplay(1));
    }
    setActiveChannel(channelId) {
        return this.dataService.client.setActiveChannel(channelId).pipe(map(({ setActiveChannel }) => setActiveChannel), tap(userStatus => {
            const activeChannel = userStatus.channels.find(c => c.id === channelId);
            if (activeChannel) {
                this.localStorageService.set('activeChannelToken', activeChannel.token);
                this.permissionsService.setCurrentUserPermissions(activeChannel.permissions);
            }
        }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: ChannelService, deps: [{ token: i1.DataService }, { token: i2.LocalStorageService }, { token: i3.PermissionsService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: ChannelService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: ChannelService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [{ type: i1.DataService }, { type: i2.LocalStorageService }, { type: i3.PermissionsService }] });
