// @vendure/admin-ui/esm2022/core/shared/components/ui-extension-point/ui-extension-point.component.mjs
import { ChangeDetectionStrategy, Component, HostBinding, Input, isDevMode, ViewChild, } from '@angular/core';
import { CodeJar } from 'codejar';
import { tap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "../../../data/providers/data.service";
import * as i2 from "@clr/angular";
import * as i3 from "@angular/common";
import * as i4 from "../dropdown/dropdown.component";
import * as i5 from "../dropdown/dropdown-menu.component";
import * as i6 from "../dropdown/dropdown-trigger.directive";
export class UiExtensionPointComponent {
    constructor(dataService) {
        this.dataService = dataService;
        this.display = 'inline-block';
        this.isDevMode = isDevMode();
    }
    getCodeTemplate(api) {
        return codeTemplates[api](this.locationId, this.metadata).trim();
    }
    ngOnInit() {
        this.display$ = this.dataService.client
            .uiState()
            .mapStream(({ uiState }) => uiState.displayUiExtensionPoints)
            .pipe(tap(display => {
            if (display) {
                setTimeout(() => {
                    const highlight = (editor) => {
                        const code = editor.textContent ?? '';
                        editor.innerHTML = highlightTsCode(this.getCodeTemplate(this.api));
                    };
                    this.editorElementRef.nativeElement.contentEditable = 'false';
                    this.jar = CodeJar(this.editorElementRef.nativeElement, highlight);
                    this.jar.updateCode(this.getCodeTemplate(this.api));
                });
            }
        }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: UiExtensionPointComponent, deps: [{ token: i1.DataService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: UiExtensionPointComponent, selector: "vdr-ui-extension-point", inputs: { locationId: "locationId", metadata: "metadata", topPx: "topPx", leftPx: "leftPx", display: "display", api: "api" }, host: { properties: { "style.display": "this.display" } }, viewQueries: [{ propertyName: "editorElementRef", first: true, predicate: ["editor"], descendants: true }, { propertyName: "dropdownComponent", first: true, predicate: ["dropdownComponent"], descendants: true }], ngImport: i0, template: "<div [class.highlight]=\"isDevMode && (display$ | async)\" class=\"wrapper\" [style.display]=\"display\">\r\n    <vdr-dropdown *ngIf=\"isDevMode && (display$ | async)\" #dropdownComponent>\r\n        <button class=\"extension-point-info-trigger\"\r\n                [style.top.px]=\"topPx ?? 0\"\r\n                [style.left.px]=\"leftPx ?? 0\"\r\n                vdrDropdownTrigger>\r\n            <clr-icon shape=\"plugin\" class=\"\" size=\"12\"></clr-icon>\r\n        </button>\r\n        <vdr-dropdown-menu>\r\n            <div #editor contenteditable=\"false\" class=\"highlighted\"></div>\r\n        </vdr-dropdown-menu>\r\n    </vdr-dropdown>\r\n    <ng-content></ng-content>\r\n</div>\r\n", styles: [":host{position:relative}.wrapper{height:100%}button.extension-point-info-trigger{position:absolute;background-color:var(--color-accent-500);border-radius:50%;width:22px;height:22px;border:0;margin:0;padding:0;z-index:100;display:flex;align-items:center;justify-content:center;opacity:.8}button.extension-point-info-trigger:hover{opacity:1}button.extension-point-info-trigger clr-icon{color:#fff}.extension-info{padding:12px}pre{padding:6px;font-family:Source Code Pro,Lucida Console,Monaco,monospace;background-color:var(--color-grey-200)}::ng-deep .highlighted{margin:calc(var(--space-unit) * 2);padding:calc(var(--space-unit) * 2);border:1px solid var(--color-grey-300);border-radius:var(--border-radius);background-color:var(--color-component-bg-200);font-family:Source Code Pro,Lucida Console,Monaco,monospace;color:var(--color-json-editor-text)}::ng-deep .highlighted .ts-string{color:var(--color-accent-600)}::ng-deep .highlighted .ts-comment{color:var(--color-grey-500)}::ng-deep .highlighted .ts-keyword{color:var(--color-json-editor-key)}::ng-deep .highlighted .ts-default{color:var(--color-json-editor-text)}\n"], dependencies: [{ kind: "directive", type: i2.ClrIconCustomTag, selector: "clr-icon" }, { kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i4.DropdownComponent, selector: "vdr-dropdown", inputs: ["manualToggle"] }, { kind: "component", type: i5.DropdownMenuComponent, selector: "vdr-dropdown-menu", inputs: ["vdrPosition", "customClasses"] }, { kind: "directive", type: i6.DropdownTriggerDirective, selector: "[vdrDropdownTrigger]" }, { kind: "pipe", type: i3.AsyncPipe, name: "async" }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: UiExtensionPointComponent, decorators: [{
            type: Component,
            args: [{ selector: 'vdr-ui-extension-point', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div [class.highlight]=\"isDevMode && (display$ | async)\" class=\"wrapper\" [style.display]=\"display\">\r\n    <vdr-dropdown *ngIf=\"isDevMode && (display$ | async)\" #dropdownComponent>\r\n        <button class=\"extension-point-info-trigger\"\r\n                [style.top.px]=\"topPx ?? 0\"\r\n                [style.left.px]=\"leftPx ?? 0\"\r\n                vdrDropdownTrigger>\r\n            <clr-icon shape=\"plugin\" class=\"\" size=\"12\"></clr-icon>\r\n        </button>\r\n        <vdr-dropdown-menu>\r\n            <div #editor contenteditable=\"false\" class=\"highlighted\"></div>\r\n        </vdr-dropdown-menu>\r\n    </vdr-dropdown>\r\n    <ng-content></ng-content>\r\n</div>\r\n", styles: [":host{position:relative}.wrapper{height:100%}button.extension-point-info-trigger{position:absolute;background-color:var(--color-accent-500);border-radius:50%;width:22px;height:22px;border:0;margin:0;padding:0;z-index:100;display:flex;align-items:center;justify-content:center;opacity:.8}button.extension-point-info-trigger:hover{opacity:1}button.extension-point-info-trigger clr-icon{color:#fff}.extension-info{padding:12px}pre{padding:6px;font-family:Source Code Pro,Lucida Console,Monaco,monospace;background-color:var(--color-grey-200)}::ng-deep .highlighted{margin:calc(var(--space-unit) * 2);padding:calc(var(--space-unit) * 2);border:1px solid var(--color-grey-300);border-radius:var(--border-radius);background-color:var(--color-component-bg-200);font-family:Source Code Pro,Lucida Console,Monaco,monospace;color:var(--color-json-editor-text)}::ng-deep .highlighted .ts-string{color:var(--color-accent-600)}::ng-deep .highlighted .ts-comment{color:var(--color-grey-500)}::ng-deep .highlighted .ts-keyword{color:var(--color-json-editor-key)}::ng-deep .highlighted .ts-default{color:var(--color-json-editor-text)}\n"] }]
        }], ctorParameters: () => [{ type: i1.DataService }], propDecorators: { locationId: [{
                type: Input
            }], metadata: [{
                type: Input
            }], topPx: [{
                type: Input
            }], leftPx: [{
                type: Input
            }], display: [{
                type: HostBinding,
                args: ['style.display']
            }, {
                type: Input
            }], api: [{
                type: Input
            }], editorElementRef: [{
                type: ViewChild,
                args: ['editor']
            }], dropdownComponent: [{
                type: ViewChild,
                args: ['dropdownComponent']
            }] } });
function highlightTsCode(tsCode) {
    tsCode = tsCode.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
    return tsCode.replace(
    // eslint-disable-next-line max-len
    /\b(abstract|any|as|boolean|break|case|catch|class|const|continue|default|do|else|enum|export|extends|false|finally|for|from|function|get|if|implements|import|in|instanceof|interface|is|keyof|let|module|namespace|never|new|null|number|object|of|private|protected|public|readonly|require|return|set|static|string|super|switch|symbol|this|throw|true|try|type|typeof|undefined|var|void|while|with|yield)\b|\/\/.*|\/\*[\s\S]*?\*\/|"(?:\\[\s\S]|[^\\"])*"|'[^']*'/g, (match, ...args) => {
        if (/^"/.test(match) || /^'/.test(match)) {
            return '<span class="ts-string">' + match + '</span>';
        }
        else if (/\/\/.*|\/\*[\s\S]*?\*\//.test(match)) {
            return '<span class="ts-comment">' + match + '</span>';
        }
        else if (
        // eslint-disable-next-line max-len
        /\b(abstract|any|as|boolean|break|case|catch|class|const|continue|default|do|else|enum|export|extends|false|finally|for|from|function|get|if|implements|import|in|instanceof|interface|is|keyof|let|module|namespace|never|new|null|number|object|of|private|protected|public|readonly|require|return|set|static|string|super|switch|symbol|this|throw|true|try|type|typeof|undefined|var|void|while|with|yield)\b/.test(match)) {
            return '<span class="ts-keyword">' + match + '</span>';
        }
        else {
            return '<span class="ts-default">' + match + '</span>';
        }
    });
}
const codeTemplates = {
    actionBar: locationId => `
import { addActionBarItem } from '@vendure/admin-ui/core';

export default [
  addActionBarItem({
    id: 'my-button',
    label: 'My Action',
    locationId: '${locationId}',
  }),
];`,
    actionBarDropdown: locationId => `
import { addActionBarDropdownMenuItem } from '@vendure/admin-ui/core';

export default [
  addActionBarDropdownMenuItem({
    id: 'my-dropdown-item',
    label: 'My Action',
    locationId: '${locationId}',
  }),
];`,
    navMenu: locationId => `
import { addNavMenuSection } from '@vendure/admin-ui/core';

export default [
  addNavMenuSection({
      id: 'my-menu-item',
      label: 'My Menu Item',
      items: [{
          // ...
      }],
    },
    '${locationId}',
  ),
];`,
    detailComponent: locationId => `
import { registerCustomDetailComponent } from '@vendure/admin-ui/core';

export default [
  registerCustomDetailComponent({
    locationId: '${locationId}',
    component: MyCustomComponent,
  }),
];`,
    dataTable: (locationId, metadata) => `
import { registerDataTableComponent } from '@vendure/admin-ui/core';

export default [
  registerDataTableComponent({
    tableId: '${locationId}',
    columnId: '${metadata}',
    component: MyCustomComponent,
  }),
];`,
};
