// @vendure/admin-ui/esm2022/core/shared/pipes/locale-currency.pipe.mjs
import { Optional, Pipe } from '@angular/core';
import { LocaleBasePipe } from './locale-base.pipe';
import * as i0 from "@angular/core";
import * as i1 from "../../providers/currency/currency.service";
import * as i2 from "../../data/providers/data.service";
/**
 * @description
 * Formats a Vendure monetary value (in cents) into the correct format for the configured currency and display
 * locale.
 *
 * @example
 * ```HTML
 * {{ variant.priceWithTax | localeCurrency }}
 * ```
 *
 * @docsCategory pipes
 */
export class LocaleCurrencyPipe extends LocaleBasePipe {
    constructor(currencyService, dataService, changeDetectorRef) {
        super(dataService, changeDetectorRef);
        this.currencyService = currencyService;
    }
    transform(value, ...args) {
        const [currencyCode, locale] = args;
        if (typeof value === 'number') {
            const activeLocale = this.getActiveLocale(locale);
            const majorUnits = this.currencyService.toMajorUnits(value);
            try {
                return new Intl.NumberFormat(activeLocale, {
                    style: 'currency',
                    currency: currencyCode,
                    minimumFractionDigits: this.currencyService.precision,
                    maximumFractionDigits: this.currencyService.precision,
                }).format(majorUnits);
            }
            catch (e) {
                return majorUnits.toFixed(this.currencyService.precision);
            }
        }
        return value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: LocaleCurrencyPipe, deps: [{ token: i1.CurrencyService }, { token: i2.DataService, optional: true }, { token: i0.ChangeDetectorRef, optional: true }], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "17.2.4", ngImport: i0, type: LocaleCurrencyPipe, name: "localeCurrency", pure: false }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: LocaleCurrencyPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'localeCurrency',
                    pure: false,
                }]
        }], ctorParameters: () => [{ type: i1.CurrencyService }, { type: i2.DataService, decorators: [{
                    type: Optional
                }] }, { type: i0.ChangeDetectorRef, decorators: [{
                    type: Optional
                }] }] });
