// @vendure/admin-ui/esm2022/core/shared/components/action-bar-items/action-bar-base.component.mjs
import { Directive, HostBinding, inject, Injector, Input } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { BehaviorSubject, of, switchMap } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { DataService } from '../../../data/providers/data.service';
import { NavBuilderService } from '../../../providers/nav-builder/nav-builder.service';
import { NotificationService } from '../../../providers/notification/notification.service';
import * as i0 from "@angular/core";
export class ActionBarBaseComponent {
    constructor() {
        this.buttonStates = {};
        this.locationId$ = new BehaviorSubject('');
        this.navBuilderService = inject(NavBuilderService);
        this.route = inject(ActivatedRoute);
        this.dataService = inject(DataService);
        this.notificationService = inject(NotificationService);
        this.injector = inject(Injector);
    }
    ngOnChanges(changes) {
        if ('locationId' in changes) {
            this.locationId$.next(changes['locationId'].currentValue);
        }
    }
    handleClick(event, item) {
        if (typeof item.onClick === 'function') {
            item.onClick(event, this.createContext());
        }
    }
    getRouterLink(item) {
        return this.navBuilderService.getRouterLink({ routerLink: item.routerLink, context: this.createContext() }, this.route);
    }
    buildButtonStates(items) {
        const context = this.createContext();
        const defaultState = {
            disabled: false,
            visible: true,
        };
        for (const item of items) {
            const buttonState$ = typeof item.buttonState === 'function'
                ? item.buttonState(context).pipe(map(result => result ?? defaultState), catchError(() => of(defaultState)))
                : of(defaultState);
            this.buttonStates[item.id] = buttonState$;
        }
    }
    createContext() {
        return {
            route: this.route,
            injector: this.injector,
            dataService: this.dataService,
            notificationService: this.notificationService,
            entity$: this.route.data.pipe(switchMap(data => {
                if (data.detail?.entity) {
                    return data.detail.entity;
                }
                else {
                    return of(undefined);
                }
            })),
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: ActionBarBaseComponent, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: ActionBarBaseComponent, inputs: { locationId: "locationId" }, host: { properties: { "attr.data-location-id": "this.locationId" } }, usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: ActionBarBaseComponent, decorators: [{
            type: Directive
        }], propDecorators: { locationId: [{
                type: HostBinding,
                args: ['attr.data-location-id']
            }, {
                type: Input
            }] } });
